src/Analyser/MutatingScope.php

```diff
@@ -4428,6 +4428,10 @@ public function addTypeToExpression(Expr $expr, Type $type): self
 
 		if ($originalExprType->equals($nativeType)) {
 			$newType = TypeCombinator::intersect($type, $originalExprType);
+			if ($newType->isConstantScalarValue()->yes() && $newType->equals($originalExprType)) {
```

Comment:
I don't get it, why only for `$newType->isConstantScalarValue()->yes()`?

And: can we make it faster for the other `return` in this method too?

Reply:
> I don't get it, why only for `$newType->isConstantScalarValue()->yes()`?

I went with constant scalars as these matched best the slow case at hand

there are types with weird equal semantics like 
- `$mixed->equals($error)` => `true`
- ObjectType
for which tests start failing if we take the fast-pass without changing the scope

> And: can we make it faster for the other return in this method too?

I can't think of a case where this would help

Reply:
These `equals` problems should be fixed, they might lead to other bugfixes as well.